File: ios/Classes/LivenessDetection/FlutterLivenessDetection.swift
//
//  FlutterLivenessDetection.swift
//  ekyc_id_flutter
//
//  Created by Socret Lee on 6/11/22.
//

import EkycID
import Foundation
import AVFoundation

public class FlutterLivenessDetection: NSObject, FlutterPlatformView, LivenessDetectionEventListener {
    let frame: CGRect
    let viewId: Int64
    var flutterScannerView: UIView?
    var scanner: LivenessDetectionView?
    var methodChannel: FlutterMethodChannel?
    var eventChannel: FlutterEventChannel?
    var eventStreamHandler: LivenessDetectionEventStreamHandler?
    
    let LIVENESS_PROMPT_TYPE_MAPPING: [String: LivenessPromptType] = [
        "BLINKING": .BLINKING,
        "LOOK_LEFT": .LOOK_LEFT,
        "LOOK_RIGHT": .LOOK_RIGHT,
    ]
    
    init(frame: CGRect, viewId: Int64, messenger: FlutterBinaryMessenger, args: Any?) {
        self.frame = frame
        self.viewId = viewId
        super.init()
        self.flutterScannerView = UIView(frame: frame)
        self.methodChannel = FlutterMethodChannel(name: "LivenessDetection_MethodChannel_" + String(viewId), binaryMessenger: messenger)
        self.eventChannel = FlutterEventChannel(name: "LivenessDetection_EventChannel_" + String(viewId), binaryMessenger: messenger)
        self.eventStreamHandler = LivenessDetectionEventStreamHandler()
        self.methodChannel!.setMethodCallHandler(self.onMethodCall)
        self.eventChannel!.setStreamHandler(self.eventStreamHandler)
    }
    
    public func view() -> UIView {
        return self.flutterScannerView!
    }
    
    private func start(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let args = call.arguments as! [String: Any]
        let prompts = args["prompts"] as! [String]
        let promptTimerCountDownSec = args["promptTimerCountDownSec"] as! Int
        self.scanner = LivenessDetectionView(frame: self.flutterScannerView!.frame)
        self.scanner!.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        self.scanner!.addListener(self)
        self.scanner!.start(
            options: LivenessDetectionOptions(
                cameraOptions: LivenessDetectionCameraOptions(
                    prompts: prompts.map { e in
                        LIVENESS_PROMPT_TYPE_MAPPING[e]!
                    },
                    promptTimerCountDownSec: promptTimerCountDownSec
                )
            )
        )
        
        self.flutterScannerView!.addSubview(self.scanner!)
        result(true)
    }
    
    
    private func dispose(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        if (self.scanner != nil) {
            self.scanner!.stop()
            self.scanner = nil
        }
        result(true)
    }
    
    private func nextImage(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        self.scanner!.nextImage()
        result(true)
    }
    
    private func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) -> Void {
        switch (call.method) {
        case "start":
            do {
                try self.start(call: call, result: result)
            } catch {
                result(FlutterError(code: "initialize Error", message: nil, details: nil))
            }
            break
        case "dispose":
            do {
                try self.dispose(call: call, result: result)
            } catch {
                result(FlutterError(code: "dispose Error", message: nil, details: nil))
            }
            break
        case "nextImage":
            do {
                try self.nextImage(call: call, result: result)
            } catch {
                result(FlutterError(code: "dispose Error", message: nil, details: nil))
            }
            break
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }
    
    public func onInitialized() {
        eventStreamHandler?.sendOnInitializedEventToFlutter()
    }
    
    public func onFocusChanged(_ isFocusing: Bool) {
        eventStreamHandler?.sendOnFocusChangedEventToFlutter(isFocusing)
    }
    
    public func onProgressChanged(_ progress: Float) {
        eventStreamHandler?.sendOnProgressChangedEventToFlutter(progress)
    }
    
    public func onCountDownChanged(current: Int, max: Int) {
        eventStreamHandler?.sendOnCountDownChangedEventToFlutter(
            current: current,
            max: max
        )
    }
    
    public func onFrameStatusChanged(_ frameStatus: EkycID.FrameStatus) {
        eventStreamHandler?.sendOnFrameStatusChangedEventToFlutter(frameStatus)
    }
    
    public func onActivePromptChanged(_ activePrompt: EkycID.LivenessPromptType?) {
        eventStreamHandler?.sendOnActivePromptChangedToFlutter(activePrompt)
    }
    
    public func onLivenessTestCompleted(_ result: EkycID.LivenessDetectionResult) {
        eventStreamHandler?.sendOnLivenessTestCompletedEventToFlutter(result)
    }
    
    class LivenessDetectionEventStreamHandler: NSObject, FlutterStreamHandler {
        var events: FlutterEventSink?
        
        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            self.events = events
            return nil
        }
        
        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            return nil
        }
        
        func sendOnInitializedEventToFlutter() {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onInitialized"
                    self.events!(event)
                }
            }
        }

        func sendOnCountDownChangedEventToFlutter(current: Int, max: Int) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onCountDownChanged"
                    var values = [String: Any?]()
                    values["current"] = current
                    values["max"] = max
                    event["values"] = values
                    self.events!(event)
                }
            }
        }
        
        func sendOnFocusChangedEventToFlutter(_ isFocusing: Bool) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onFocusChanged"
                    event["value"] = isFocusing
                    self.events!(event)
                }
            }
        }
        
        func sendOnProgressChangedEventToFlutter(_ progress: Float) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onProgressChanged"
                    event["value"] = progress
                    self.events!(event)
                }
            }
        }
        
        func sendOnFrameStatusChangedEventToFlutter(_ frameStatus: EkycID.FrameStatus) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onFrameStatusChanged"
                    event["values"] = "\(frameStatus)"
                    self.events!(event)
                }
            }
        }
        
        func sendOnActivePromptChangedToFlutter(_ activePrompt: EkycID.LivenessPromptType?) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onActivePromptChanged"
                    if (activePrompt != nil) {
                        event["values"] = "\(activePrompt)"
                    }
                    self.events!(event)
                }
            }
        }
        
        func sendOnLivenessTestCompletedEventToFlutter(_ result: LivenessDetectionResult) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onLivenessTestCompleted"
                    event["values"] = result.toFlutterMap()
                    self.events!(event)
                }
            }
        }
    }
}

extension LivenessPrompt {
    func toFlutterMap() -> [String: Any?] {
        var values = [String: Any?]()
        values["prompt"] = "\(self.prompt)"
        values["success"] = self.success
        return values
    }
}

extension LivenessFace {
    func toFlutterMap() -> [String: Any?] {
        var values = [String: Any?]()
        values["image"] = FlutterStandardTypedData(bytes: self.image!.jpegData(compressionQuality: 0.8)!)
        values["leftEyeOpenProbability"] = self.leftEyeOpenProbability
        values["rightEyeOpenProbability"] = self.rightEyeOpenProbability
        values["headEulerAngleX"] = self.headEulerAngleX
        values["headEulerAngleY"] = self.headEulerAngleY
        values["headEulerAngleZ"] = self.headEulerAngleZ
        
        if (self.headDirection != nil) {
            values["headDirection"] = "\(self.headDirection!)"
        } else {
            values["headDirection"] = nil
        }
        
        if (self.eyesStatus != nil) {
            values["eyesStatus"] = "\(self.eyesStatus!)"
        } else {
            values["eyesStatus"] = nil
        }
        return values
    }
}

extension LivenessDetectionResult {
    func toFlutterMap() -> [String: Any?] {
        var values = [String: Any?]()
        if (self.frontFace != nil) {
            values["frontFace"] = self.frontFace!.toFlutterMap()
        } else {
            values["frontFace"] = nil
        }
        
        if (self.leftFace != nil) {
            values["leftFace"] = self.leftFace!.toFlutterMap()
        } else {
            values["leftFace"] = nil
        }
        
        if (self.rightFace != nil) {
            values["rightFace"] = self.rightFace!.toFlutterMap()
        } else {
            values["rightFace"] = nil
        }
        
        values["prompts"] = self.prompts.map {e -> [String:Any?] in
            var v = [String:Any?]()
            v["prompt"]="\(e.prompt)"
            v["success"]=e.success
            return v
        }
        
        return values
    }
}

File: android/src/main/kotlin/com/ekycsolutions/ekyc_id_flutter/LivenessDetection/FlutterLivenessDetection.kt
package com.ekycsolutions.ekyc_id_flutter.LivenessDetection

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import com.ekycsolutions.ekyc_id_flutter.R
import com.ekycsolutions.ekycid.core.models.FrameStatus
import com.ekycsolutions.ekycid.livenessdetection.LivenessDetectionEventListener
import com.ekycsolutions.ekycid.livenessdetection.LivenessDetectionResult
import com.ekycsolutions.ekycid.livenessdetection.LivenessDetectionView
import com.ekycsolutions.ekycid.livenessdetection.cameraview.LivenessFace
import com.ekycsolutions.ekycid.livenessdetection.cameraview.LivenessPromptType

import java.io.ByteArrayOutputStream
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import io.flutter.embedding.engine.plugins.FlutterPlugin

class FlutterLivenessDetection(
    private var binding: FlutterPlugin.FlutterPluginBinding,
    private var context: Context,
    private val viewId: Int
) : PlatformView, MethodChannel.MethodCallHandler, LivenessDetectionEventListener {
    private var scanner: LivenessDetectionView? = null

    private var scannerView: LinearLayout = LinearLayout(context)

    private val methodChannel: MethodChannel =
        MethodChannel(binding.binaryMessenger, "LivenessDetection_MethodChannel_$viewId")
    private val eventChannel: EventChannel =
        EventChannel(binding.binaryMessenger, "LivenessDetection_EventChannel_$viewId")
    private val eventStreamHandler = LivenessDetectionEventStreamHandler(context)

//    private val LIVENESS_PROMPT_TYPE_MAPPING: HashMap<String, LivenessPromptType> = hashMapOf(
//        "BLINKING" to LivenessPromptType.BLINKING,
//        "LOOK_LEFT" to LivenessPromptType.LOOK_LEFT,
//        "LOOK_RIGHT" to LivenessPromptType.LOOK_RIGHT
//    )

    init {
        this.methodChannel.setMethodCallHandler(this)
        this.eventChannel.setStreamHandler(eventStreamHandler)
    }


    override fun getView(): View {
        return scannerView
    }

    override fun dispose() {

    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> {
                start(call, result)
            }
            "dispose" -> {
                disposeFlutter(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun start(call: MethodCall, result: MethodChannel.Result) {
        try {
            val args = call.arguments as HashMap<*, *>
//            val prompts = args["prompts"] as ArrayList<String>
//            val prompts = args["prompts"] as ArrayList<HashMap<String,Any>>
            val language = args["language"] as String
            val promptTimerCountDownSec = args["promptTimerCountDownSec"] as Int
//            this.cameraView!!.setOptions(
//                LivenessDetectionOptions(
//                    ArrayList(prompts.map {
//                        LIVENESS_PROMPT_TYPE_MAPPING[it]!!
//                    }),
//                    promptTimerCountDownSec
//                )
//            )
            this.scanner!!.addListener(this)
            this.scanner!!.start()
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    private fun disposeFlutter(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (this.scanner != null) {
                this.scanner!!.stop()
                this.scanner = null
            }
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    private fun nextImage(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (scanner != null) {
            }
//            this.cameraView!!.nextImage()
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

//    override fun onInitialize() {
//        this.eventStreamHandler.sendOnInitializedEventToFlutter()
//    }

    override fun onActivePromptChanged(activePrompt: LivenessPromptType?) {
        eventStreamHandler.sendOnActivePromptChangedEventToFlutter(activePrompt)
    }

    override fun onCountDownChanged(current: Int, max: Int) {
        eventStreamHandler.sendOnCountDownChangedEventToFlutter(current, max)
    }

    override fun onFocusChanged(isFocusing: Boolean) {
        eventStreamHandler.sendOnFocusChangedEventToFlutter(isFocusing)
    }

    override fun onFrameStatusChanged(frameStatus: FrameStatus) {
        eventStreamHandler.sendOnFrameStatusChangedEventToFlutter(frameStatus)
    }

    override fun onLivenessTestCompleted(result: LivenessDetectionResult) {
        eventStreamHandler.sendOnLivenessTestCompletedEventToFlutter(result)
    }

    override fun onProgressChanged(progress: Float) {
        eventStreamHandler.sendOnProgressChangedEventToFlutter(progress)
    }

    class LivenessDetectionEventStreamHandler(private var context: Context) :
        EventChannel.StreamHandler {
        private var events: EventChannel.EventSink? = null

        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            this.events = events
        }

        override fun onCancel(arguments: Any?) {
            this.events = null
        }

        fun sendOnActivePromptChangedEventToFlutter(activePrompt: LivenessPromptType?) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onActivePromptChanged"
                    event["values"] = "LivenessPromptType.$activePrompt"
                    events!!.success(event)
                }
            }
        }


        fun sendOnFocusChangedEventToFlutter(isFocusing: Boolean) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onFocusChanged"
                    event["values"] = isFocusing
                    events!!.success(event)
                }
            }
        }

        fun sendOnFrameStatusChangedEventToFlutter(frameStatus: FrameStatus) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onFrameStatusChanged"
                    event["values"] = frameStatus.name
                    events!!.success(event)
                }
            }
        }

        fun sendOnProgressChangedEventToFlutter(progress: Float) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onProgressChanged"
                    event["values"] = progress
                    events!!.success(event)
                }
            }
        }

        fun sendOnLivenessTestCompletedEventToFlutter(result: LivenessDetectionResult) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    val values = HashMap<String, Any?>()

                    var promptString = arrayListOf<HashMap<String,Any?>>()

                    for (p in result.prompts) {
                        val v = HashMap<String, Any?>()
                        v["prompt"] = p.prompt.name
                        v["success"] = p.success
                        promptString.add(v)
                    }

                    values["prompts"] = promptString
                    values["frontFace"] = livenessFaceToFlutterMap(result.frontFace)
                    values["leftFace"] = livenessFaceToFlutterMap(result.leftFace)
                    values["rightFace"] = livenessFaceToFlutterMap(result.rightFace)
                    event["type"] = "onLivenessTestCompleted"
                    event["values"] = values
                    events!!.success(event)
                }
            }
        }

        fun sendOnCountDownChangedEventToFlutter(current: Int, max: Int) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onCountDownChanged"
                    val values = HashMap<String, Any?>()
                    values["current"] = current
                    values["max"] = max
                    event["values"] = values
                    events!!.success(event)
                }
            }
        }

        private fun livenessFaceToFlutterMap(livenessFace: LivenessFace?): HashMap<String, Any?>? {

            if (livenessFace == null) {
                return null
            }

            val values = HashMap<String, Any?>()
            values["image"] =
                if (livenessFace.image != null) bitmapToFlutterByteArray(livenessFace.image!!) else null
            values["leftEyeOpenProbability"] = livenessFace.leftEyeOpenProbability
            values["rightEyeOpenProbability"] = livenessFace.rightEyeOpenProbability
            values["headEulerAngleX"] = livenessFace.headEulerAngleX
            values["headEulerAngleY"] = livenessFace.headEulerAngleY
            values["headEulerAngleZ"] = livenessFace.headEulerAngleZ

            if (livenessFace.headDirection != null) {
                values["headDirection"] = livenessFace.headDirection!!.name
            } else {
                values["headDirection"] = null
            }

            if (livenessFace.eyesStatus != null) {
                values["eyesStatus"] = livenessFace.eyesStatus!!.name
            } else {
                values["eyesStatus"] = null
            }

            return values
        }

        private fun bitmapToFlutterByteArray(image: Bitmap): ByteArray {
            val stream = ByteArrayOutputStream()
            image.compress(Bitmap.CompressFormat.JPEG, 90, stream)
            return stream.toByteArray()
        }
    }
}